### PR TITLE
⚠️ Refactor the ClusterClass codebase and reconciler around a scope

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR introduces a scope object for holding objects required when processing a request in Topology/ClusterReconciler

The scope object holds the info previously defined as a separated struct in topology/types.go; Additionally.
- The types ending with TopologyClass where now renamed in Blueprint thus removing some confusion with ClusterClass
- Cluster.Topology has been added to the blueprint, thus avoiding to access cluster.Current as source of information about the Topology
- Usage of blueprint.HasControlPlaneInfrastructureMachine has been made consistent
- The types ending with TopologyState where now renamed in State

### CODE REVIEW

File: controllers/topology/controller.go [hunk 1]
```diff
@@ -117,32 +118,39 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
...
 		return ctrl.Result{}, errors.Wrap(err, "error reading current state of the Cluster topology")
 	}
 
-	// Computes the desired state of the Cluster
-	desiredState, err := r.computeDesiredState(ctx, class, currentState)
+	// Computes the desired state of the Cluster and store it in the request scope.
+	scope.Desired, err = r.computeDesiredState(ctx, scope)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "error computing the desired state of the Cluster topology")
 	}
```
Comment (vincepri) [inline controllers/topology/controller.go:150]: Could we build these in the scope package instead?

File: controllers/topology/blueprint.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+// getBlueprint gets a ClusterBlueprint with the ClusterClass and the referenced templates to be used for a managed Cluster topology.
+// It also converts and patches all ObjectReferences in ClusterClass and ControlPlane to the latest apiVersion of the current contract.
+// NOTE: This function assumes that cluster.Spec.Topology.Class is set.
+func (r *ClusterReconciler) getBlueprint(ctx context.Context, s *scope.Scope) (_ *scope.ClusterBlueprint, reterr error) {
+	cluster := s.Current.Cluster
+	blueprint := &scope.ClusterBlueprint{
+		Topology:           cluster.Spec.Topology,
+		ClusterClass:       &clusterv1.ClusterClass{},
+		MachineDeployments: map[string]*scope.MachineDeploymentBlueprint{},
```
Comment (ykakarap) [inline controllers/topology/blueprint.go:38]:
```suggestion
                 ControlPlane : &scope.ControlPlaneBlueprint{},
		MachineDeployments: map[string]*scope.MachineDeploymentBlueprint{},
```
also initialize ControlPlane?

File: controllers/topology/internal/scope/blueprint.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	// BootstrapTemplate holds the bootstrap template for a MachineDeployment referenced from ClusterClass.
+	BootstrapTemplate *unstructured.Unstructured
+
+	// InfrastructureMachineTemplate holds the infrastructure machine template for a MachineDeployment referenced from ClusterClass.
+	InfrastructureMachineTemplate *unstructured.Unstructured
+}
+
+// HasControlPlaneInfrastructureMachine checks whether the clusterClass mandates the controlPlane has infrastructureMachines.
+func (b *ClusterBlueprint) HasControlPlaneInfrastructureMachine() bool {
+	return b.ClusterClass.Spec.ControlPlane.MachineInfrastructure != nil && b.ClusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil
```
Comment (ykakarap) [inline controllers/topology/internal/scope/blueprint.go:69]: would it be better to make sure that `ClusterClass` is set in the blueprint?

File: controllers/topology/controller.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/topology/controller.go:150]: This PR is already huge.
Could we reconsider this in a separated PR (I'm not fully convinced moving all the code in the scope provides concrete benefits in code organisation.
Also, there is work to be done to add to the scope all the necessary machinery for avoiding orphaned objects...

File: controllers/topology/blueprint.go (hunk 2 above)
Comment (fabriziopandini) [reply controllers/topology/blueprint.go:38]: This is currently done down in the code; we can eventually reshuffle things in a follow up

File: controllers/topology/controller.go (hunk 1 above)
Comment (vincepri) [reply controllers/topology/controller.go:150]: From a purely code perspective, this code is defining types in a package with a `New()` function that doesn't do much, and then we change the fields on the caller. In general, this isn't well contained behavior and would want to consider moving at least how we compose the scope to its respective package.

We can do this in another PR

File: controllers/topology/internal/scope/blueprint.go (hunk 3 above)
Comment (fabriziopandini) [reply controllers/topology/internal/scope/blueprint.go:69]: It is a double edged sword.
Returning false in case ClusterClass is not set will provide a false signal; returning errors makes consumption of this function less straightforward.

**Comment by dharmjit** [comment]
LGTM, I just wanted to look at the GitHub action results. Do you have some PRs in your repo with the action results?

File: controllers/topology/internal/scope/blueprint.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+package scope
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+// ClusterBlueprint holds all the objects required for computing the desired state of a managed Cluster topology,
+// including the ClusterClass and all the referenced templates.
+type ClusterBlueprint struct {
```
Comment (killianmuldoon) [inline controllers/topology/internal/scope/blueprint.go:26]: Could we plan to add some additional documentation about these types? There's  semantic overlap between Template and Blueprint that's confusing at first glance at least. The flow here of Scope -> ClusterBlueprint -> ControlPlaneBlueprint -> Template etc. is deep and could be hard to follow

Comment (killianmuldoon) [reply controllers/topology/internal/scope/blueprint.go:26]: The questions 'What is a blueprint?' and 'What is a scope' need good answers, especially if this pattern is repeated in future.

File: controllers/topology/reconcile_state.go [hunk 5]
```diff
@@ -114,47 +112,47 @@ func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, class *co
...
+			return errors.Wrapf(err, "failed to patch %s/%s", scope.Current.Cluster.GroupVersionKind(), scope.Current.Cluster.Name)
 		}
 	}
 	return nil
 }
 
 // reconcileMachineDeployments reconciles the desired state of the MachineDeployment objects.
-func (r *ClusterReconciler) reconcileMachineDeployments(ctx context.Context, current, desired *clusterTopologyState) error {
-	diff := calculateMachineDeploymentDiff(current.machineDeployments, desired.machineDeployments)
+func (r *ClusterReconciler) reconcileMachineDeployments(ctx context.Context, scope *scope.Scope) error {
```
Comment (killianmuldoon) [inline controllers/topology/reconcile_state.go:133]: Does it make sense to shorten 'scope' to 's' here? The variables being passed down to subordinate functions are very long and had to read. Shortening the variable would at least make the Current vs Desired distinction stand out and be front and center across the reconciliation functions.

Comment (fabriziopandini) [reply controllers/topology/reconcile_state.go]: Done!

File: controllers/topology/internal/scope/blueprint.go (hunk 4 above)
Comment (fabriziopandini) [reply controllers/topology/internal/scope/blueprint.go:26]: > Could we plan to add some additional documentation about these types? 

Blueprint was introduced to replace class, which in turns was creating confusion with ClusterClass.

Current definitions are:
```
// ClusterBlueprint holds all the objects required for computing the desired state of a managed Cluster topology,
// including the ClusterClass and all the referenced templates.
type ClusterBlueprint struct {
   ...
}

// Scope holds all the information to process a request in the topology/ClusterReconciler controller.
type Scope struct {
   ...
}
```

Happy to improve them if there are suggestions...

> The flow here of Scope -> ClusterBlueprint -> ControlPlaneBlueprint -> Template etc. is deep and could be hard to follow

Nested types are hard to follow, but on the other side they provide an organization in this struct, which reads as
`s.Blueprint.controlplane.template` when used in the code.

> The questions 'What is a blueprint?' and 'What is a scope' need good answers, especially if this pattern is repeated in future.

I'm not sure the blueprint is something that will become a pattern, topology/ClusterRecociler is quite different from existing reconcilers.

Scope instead is already a pattern in some providers (e.g. CAPA); also KCP has a complicated variant of it; the name is kind of debatable, given that it is a request context, but for this first iteration I stuck to what is used in the community

Comment (killianmuldoon) [reply controllers/topology/internal/scope/blueprint.go:26]: With that context scope definitely makes sense - It would be good to get a one-line answer to differentiate what a Blueprint is versus a Template (or even a Spec!) though, but it's not super important unless the patter might be repeated elsewhere.

File: controllers/topology/desired_state.go [hunk 6]
```diff
@@ -28,143 +28,164 @@ import (
...
 
 	// Compute the desired state for the Cluster object adding a reference to the
 	// InfrastructureCluster and the ControlPlane objects generated by the previous step.
-	desiredState.cluster = computeCluster(current, desiredState.infrastructureCluster, desiredState.controlPlane.object)
+	desiredState.Cluster = computeCluster(ctx, s, desiredState.InfrastructureCluster, desiredState.ControlPlane.Object)
 
-	// Compute the desired state of the MachineDeployment objects for the worker nodes.
-	if current.cluster.Spec.Topology.Workers == nil || len(current.cluster.Spec.Topology.Workers.MachineDeployments) == 0 {
+	// If required by the blueprint, compute the desired state of the MachineDeployment objects for the worker nodes.
+	if s.Blueprint.HasMachineDeployments() {
```
Comment (sbueringer) [inline controllers/topology/desired_state.go:67]: That also seems to be inverted (like the implementation). So I guess we're fine :)

**Review by sbueringer — COMMENTED** [review]
only a few nits, otherwise lgtm. I like the new naming!

**Comment by vincepri** [comment]
/retitle ⚠️ Refactor the ClusterClass codebase and reconciler around a scope

Breaking changes detected:
- `ClusterTopologyMachineDeploymentLabelName`: value changed from "cluster.x-k8s.io/topology/deployment-name" to "topology.cluster.x-k8s.io/deployment-name" <-- **This label was invalid and has been fixed**
- `ControlPlaneTopology.Replicas`: changed from *int to *int32 **This field is behind an alpha feature gate**
- `MachineDeploymentTopology.Replicas`: changed from *int to *int32 **This field is behind an alpha feature gate**

File: controllers/topology/desired_state.go (hunk 6 above)
Comment (fabriziopandini) [reply controllers/topology/desired_state.go:67]: fixed in https://github.com/kubernetes-sigs/cluster-api/pull/5128

File: controllers/topology/internal/scope/scope.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+// Scope holds all the information to process a request in the topology/ClusterReconciler controller.
+type Scope struct {
+	// Blueprint holds all the objects required for computing the desired state of a managed topology.
+	Blueprint *ClusterBlueprint
+
+	// Current holds the current state of the managed topology.
+	Current *ClusterState
+
+	// Desired holds the current state of the managed topology.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply controllers/topology/internal/scope/scope.go:31]: fixed in https://github.com/kubernetes-sigs/cluster-api/pull/5128

File: controllers/topology/internal/scope/blueprint.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// HasControlPlaneInfrastructureMachine checks whether the clusterClass mandates the controlPlane has infrastructureMachines.
+func (b *ClusterBlueprint) HasControlPlaneInfrastructureMachine() bool {
+	return b.ClusterClass.Spec.ControlPlane.MachineInfrastructure != nil && b.ClusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil
+}
+
+// HasMachineDeployments checks whether the topology has MachineDeployments.
+func (b *ClusterBlueprint) HasMachineDeployments() bool {
+	return b.Topology.Workers == nil || len(b.Topology.Workers.MachineDeployments) == 0
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply controllers/topology/internal/scope/blueprint.go:74]: fixed in https://github.com/kubernetes-sigs/cluster-api/pull/5128
